File: MRZvIS/LW1/LW1/Pipeline.cpp
#include "Pipeline.h"


void PipelineStepStruct::execute() const
{
	p_function(p_data);
}

void PipelineStruct::tick()
{
    static size_t s_steps_count = steps_count;
    static size_t s_current_input = 0;
    static size_t s_current_output = 0;

    printf("\n# TICK %u\n", current_tick);

    // move output to next steps' input
    for (size_t i = s_steps_count - 1; i > 0; i--) {
        p_steps[i].p_data = p_steps[i - 1].p_data;
    }

    // move input data for first step
    if (s_current_input < input_count) {
        p_steps[0].p_data = p_input[s_current_input];
        p_input[s_current_input] = NULL;
        s_current_input++;
    }
    else {
        p_steps[0].p_data = NULL; // if input array is exhausted
    }

    // invoke pipeline steps
    for (size_t i = 0; i < s_steps_count; i++) {
        printf("## STEP %d:\n", i);

        if (p_steps[i].p_data != NULL) {
            p_steps[i].execute();
        }
        else {
            printf("no input data\n");
        }
    }

    current_tick++;
}

File: MRZvIS/LW1/LW1/Pipeline.cs
// Лабораторная работа №1 по дисциплине МРЗвИС
// Вариант 1: алгоритм вычисления произведения пары 4-разрядных чисел умножением с младших разрядов со сдвигом множимого (частичного произведения) влево
// Выполнил студент группы 221701 БГУИР Робилко Тимур Маркович
//
// Файл, содержащий класс конвейера
//
// Источники:
// - Формальные модели обработки информации и параллельные модели решения задач : учеб.-метод. пособие / В. П. Ивашенко. – Минск : БГУИР, 2020


namespace LW1
{
    public class Pipeline
    {
        public delegate void PipelineStageFunction(MultiplicationTriple triple);
        private class PipelineStage(int index, PipelineStageFunction func)
        {
            public int Index { get; init; } = index;
            public PipelineStageFunction Function { get; init; } = func;
            public MultiplicationTriple? Data;

            public void Execute()
            {
                if (Data is not null)
                {
                    Function.Invoke(Data);
                }
            }
        }

        private readonly List<PipelineStage> Stages = [];

        public Queue<MultiplicationTriple> Input { get; private set; } = [];
        public Queue<MultiplicationTriple> Output { get; private set; } = [];
        public int CurrentTick { get; private set; }

        public void AddStage(PipelineStageFunction func)
        {
            Stages.Add(new PipelineStage(Stages.Count, func));
        }

        public bool Tick()
        {
            if(Stages.Count == 0)
            {
                CurrentTick++;
                return false;
            }

            Debug.Log($"\nТАКТ {CurrentTick++}\nВходная очередь:");
            foreach (var triple in Input)
            {
                Debug.Log($"{triple.Multiplicand}\t({(uint)triple.Multiplicand}),\t{triple.Factor}\t({(uint)triple.Factor})");
            }
            Debug.Log();

            foreach (var stage in Stages)
            {
                Debug.Log($"Этап {stage.Index}");
                stage.Execute();
                Debug.Log();
            }

            var lastData = Stages.Last().Data;
            if (lastData is not null)
            {
                Output.Enqueue(lastData);
            }

            for (int i = Stages.Count - 1; i > 0; i--)
            {
                Stages[i].Data = Stages[i - 1].Data;
            }

            Stages.First().Data = Input.Count > 0 ? Input.Dequeue() : null;

            Debug.Log("Выход:");
            foreach (var triple in Output)
            {
                Debug.Log($"{triple.PartialSum}\t({(uint)triple.PartialSum})");
            }
            Debug.Log();

            var nonEmptyStagesCount = Stages.Count(stage => stage.Data is not null);
            return nonEmptyStagesCount > 0 || CurrentTick == 1;
        }
    };
}